// oauth2-ts/dist/lib/models/token-model.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const invalid_argument_error_1 = require("../errors/invalid-argument-error");
const modelAttributes = [
    'accessToken',
    'accessTokenExpiresAt',
    'client',
    'refreshToken',
    'refreshTokenExpiresAt',
    'scope',
    'user',
];
class TokenModel {
    constructor(data = {}, options = {}) {
        if (!data.accessToken) {
            throw new invalid_argument_error_1.InvalidArgumentError('Missing parameter: `accessToken`');
        }
        if (!data.client) {
            throw new invalid_argument_error_1.InvalidArgumentError('Missing parameter: `client`');
        }
        if (!data.user) {
            throw new invalid_argument_error_1.InvalidArgumentError('Missing parameter: `user`');
        }
        if (data.accessTokenExpiresAt &&
            !(data.accessTokenExpiresAt instanceof Date)) {
            throw new invalid_argument_error_1.InvalidArgumentError('Invalid parameter: `accessTokenExpiresAt`');
        }
        if (data.refreshTokenExpiresAt &&
            !(data.refreshTokenExpiresAt instanceof Date)) {
            throw new invalid_argument_error_1.InvalidArgumentError('Invalid parameter: `refreshTokenExpiresAt`');
        }
        this.accessToken = data.accessToken;
        this.accessTokenExpiresAt = data.accessTokenExpiresAt;
        this.client = data.client;
        this.refreshToken = data.refreshToken;
        this.refreshTokenExpiresAt = data.refreshTokenExpiresAt;
        this.scope = data.scope;
        this.user = data.user;
        if (options && options.allowExtendedTokenAttributes) {
            this.customAttributes = {};
            for (const key in data) {
                if (data.hasOwnProperty(key) && modelAttributes.indexOf(key) < 0) {
                    this.customAttributes[key] = data[key];
                }
            }
        }
        const msInS = 1000;
        if (this.accessTokenExpiresAt) {
            this.accessTokenLifetime = Math.floor((this.accessTokenExpiresAt.getTime() - new Date().getTime()) / msInS);
        }
    }
}
exports.TokenModel = TokenModel;
